bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -195,23 +196,38 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 		return ctrl.Result{}, err
 	}
 
+	// Attempt to Patch the KubeadmConfig object and status after each reconciliation if no error occurs.
+	defer func() {
+		// always update the readyCondition; the summary is represented using the "1 of x completed" notation.
+		conditions.SetSummary(config, conditions.WithStepCounter(1))
+
+		if err := patchHelper.Patch(ctx, config); err != nil {
+			log.Error(rerr, "failed to patch config")
+			if rerr == nil {
+				rerr = err
+			}
+		}
+	}()
+
 	switch {
 	// Wait for the infrastructure to be ready.
 	case !cluster.Status.InfrastructureReady:
 		log.Info("Cluster infrastructure is not ready, waiting")
+		conditions.MarkFalse(config, bootstrapv1.DataSecretGeneratedCondition, bootstrapv1.WaitingForClusterInfrastructureReason, clusterv1.ConditionSeverityInfo, "")
```

Comment:
Out of curiosity, what would happen here if `cluster.Status.InfrastructureReady` flipped from `True` to `False` after we already generated the bootstrap data secret? I would expect that we wouldn't change the condition here because the secret is already generated. Should this only set the condition as false if the condition doesn't already exist or is false and the reason differs?

Reply:
If it is possible that `cluster.Status.InfrastructureReady` could flip back to false, then we should wrap this in an `if config.Status.DataSecretName == nil {}`. 
Do you expect this case can actually happen?

Reply:
The naive case I can think of is after a "pivot" operation, if the KubeadmConfig is reconciled prior to the associated Cluster resource after unpausing.

Reply:
I don't think checking our own status would be sufficient here because we haven't reconciled the status at this point.

Reply:
If the `InfrastructureReady` on Cluster flips after the secret has been generated, there shouldn't be anything for CABPK to do at that point.

As a general guideline, we should be able to rely on external object status (esp. conditions), in this case Cluster.